src/Microsoft.ML.StaticPipe/KMeansStatic.cs

```diff
@@ -30,19 +31,62 @@ public static class KMeansClusteringExtensions
         public static (Vector<float> score, Key<uint> predictedLabel) KMeans(this ClusteringContext.ClusteringTrainers ctx,
            Vector<float> features, Scalar<float> weights = null,
            int clustersCount = KMeansPlusPlusTrainer.Defaults.K,
-           Action<KMeansPlusPlusTrainer.Arguments> advancedSettings = null,
            Action<KMeansModelParameters> onFit = null)
         {
             Contracts.CheckValue(features, nameof(features));
             Contracts.CheckValueOrNull(weights);
             Contracts.CheckParam(clustersCount > 1, nameof(clustersCount), "If provided, must be greater than 1.");
             Contracts.CheckValueOrNull(onFit);
-            Contracts.CheckValueOrNull(advancedSettings);
 
             var rec = new TrainerEstimatorReconciler.Clustering(
             (env, featuresName, weightsName) =>
             {
-                var trainer = new KMeansPlusPlusTrainer(env, featuresName, clustersCount, weightsName, advancedSettings);
+                var options = new KMeansPlusPlusTrainer.Options
+                {
+                    FeatureColumn = featuresName,
+                    K = clustersCount,
+                    WeightColumn = weightsName != null ? Optional<string>.Explicit(weightsName) : Optional<string>.Implicit(DefaultColumnNames.Weight)
+                };
+
+                var trainer = new KMeansPlusPlusTrainer(env, options);
+
+                if (onFit != null)
+                    return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
+                else
+                    return trainer;
+            }, features, weights);
+
+            return rec.Output;
+        }
+
+        /// <summary>
+        /// KMeans <see cref="ClusteringContext"/> extension method.
+        /// </summary>
+        /// <param name="ctx">The regression context trainer object.</param>
+        /// <param name="features">The features, or independent variables.</param>
+        /// <param name="weights">The optional example weights.</param>
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Vector<float> score, Key<uint> predictedLabel) KMeans(this ClusteringContext.ClusteringTrainers ctx,
+           Vector<float> features, Scalar<float> weights = null,
+           KMeansPlusPlusTrainer.Options options = null,
```

Comment:
the API pattern I am following is that both `weights`  and `options` are required parameters  for the `Options` API

Reply:
I have a different question, why do we want to the features, and the weight in this second extension; can we just have it be options, and provide everything in there?

---
In reply to: [248121755](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248121755) [](ancestors = 248121755)

Reply:
Actually, quoting from: [https://github.com/dotnet/machinelearning/issues/1798](https://github.com/dotnet/machinelearning/issues/1798)

"It is good to have the convenience for the simple arguments, however, if we have both simple and advanced settings, we should not mix them but have instead two distinct constructors/extension methods. (E.g., in the above, we would have two methods, one that took the advanced options.) To do otherwise is to invite confusion about which "wins" if we have the setting set in both.

Note that phase setting "set in both," which suggests that these settings object should retain the "simpler" settings in them. This reinforces feedback elsewhere as seen here."


Seems like the second extension should not have the label, features, weights in the signature; but just Options. 

---
In reply to: [248157402](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248157402) [](ancestors = 248157402,248121755)